### PR TITLE
Modifies CMake to always use vcpkg, optionally have an install target…

### LINKED ISSUE #178: Create CMake install target for `librellic.a`

### PR DESCRIPTION
**Pull request by pgoodman**
Fixes #178

### CODE REVIEW

**Comment by frabert** [comment]
It would probably be good to also modify the way Python scripts are called, e.g.

```cmake
add_test(NAME test_roundtrip_rebuild
  COMMAND scripts/roundtrip.py $<TARGET_FILE:${RELLIC_DECOMP}> tests/tools/decomp/ "${CLANG_PATH}"
  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
)
```

should instead be

```cmake
find_package(Python3 COMPONENTS Interpreter REQUIRED)
add_test(NAME test_roundtrip_rebuild
  COMMAND ${Python3_EXECUTABLE} scripts/roundtrip.py $<TARGET_FILE:${RELLIC_DECOMP}> tests/tools/decomp/ "${CLANG_PATH}"
  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
)
```

File: README.md [hunk 1]
```diff
@@ -85,7 +85,24 @@ To try out Rellic you can do the following, given a LLVM bitcode file of your ch
...
+
+### On macOS
+
+Make sure to have the latest release of cxx-common for LLVM 12. Then, build with
+
+```shell
+cmake \
+  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
+  -DVCPKG_ROOT=/path/to/vcpkg \
+  -DVCPKG_TARGET_TRIPLET=x64-osx \
```
Comment (ekilmer) [inline README.md:99]: I think this should be 
```suggestion
  -DVCPKG_TARGET_TRIPLET=x64-osx-rel \
```
as a default for using the "Release" versions, which is what people will download from cxx-common

File: rellicConfig.cmake.in [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# All rights reserved.
+#
+# This source code is licensed in accordance with the terms specified in
+# the LICENSE file found in the root directory of this source tree.
+#
+
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/@PROJECT_NAME@Targets.cmake")
+check_required_components("@PROJECT_NAME@")
```
Comment (ekilmer) [inline rellicConfig.cmake.in:12]: This file needs to include `find_dependency` calls to LLVM, Clang, glog, gflags, and all of the other public third party interfaces in the rellic library(s).

https://cmake.org/cmake/help/latest/module/CMakeFindDependencyMacro.html

File: CMakeLists.txt [hunk 3]
```diff
@@ -46,65 +40,31 @@ endif()
...
-if (NOT gflags_FOUND)
-  find_package(gflags REQUIRED)
+# LLVM uses `-lZ3` somewhere, so this is a good way to sneak it in.
+# NOTE: Recent versions of cxx-common vcpkg llvm dependencies fix this
+if(NOT Z3_FOUND)
+    find_package(Z3 4.8 CONFIG REQUIRED)
+    if (NOT TARGET Z3)
+        add_library(Z3 INTERFACE)
+    endif()
+    target_link_libraries(Z3 INTERFACE z3::libz3)
```
Comment (ekilmer) [inline CMakeLists.txt:63]: What was wrong with the logic that was here before for setting up/fixing the Z3 linkage?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,31 +1,25 @@
...
-endif ()
-
-if (DEFINED VCPKG_ROOT)
-  message(WARNING "Building with VCPKG")
-  include(cmake/vcpkg_helper.cmake)
-else()
-  message(WARNING "Running the standard CMake build; if this is a mistake, reconfigure again and pass `-DVCPKG_ROOT=/path/to/vcpkg`")
-  set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
-endif()
+include("cmake/vcpkg_helper.cmake")
```
Comment (ekilmer) [inline CMakeLists.txt:9]: What's the use of removing the option to build in a different way? I felt like the warning messages and README were good enough for notifying users that they are building one way or another and that vcpkg was the supported way.

File: CMakeLists.txt [hunk 5]
```diff
@@ -46,65 +40,31 @@ endif()
...
-# Potentially fixup bad LLVM z3 linkage
-get_target_property(LLVMSupport_LIBS LLVMSupport INTERFACE_LINK_LIBRARIES)
-list(FIND LLVMSupport_LIBS "Z3" _index)
-if (${_index} GREATER -1)
-  list(REMOVE_ITEM LLVMSupport_LIBS Z3)
-  list(APPEND LLVMSupport_LIBS z3::libz3)
-  set_target_properties(LLVMSupport PROPERTIES
-    INTERFACE_LINK_LIBRARIES "${LLVMSupport_LIBS}")
-endif()
+set(_VCPKG_SHARE_DIR "${VCPKG_ROOT}/installed/${VCPKG_TARGET_TRIPLET}/${CMAKE_INSTALL_DATADIR}")
```
Comment (ekilmer) [inline CMakeLists.txt:50]: Vcpkg sets this up already. This should not be required for correct use of vcpkg packages.
```suggestion
```

File: cmake/llvm.cmake [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    endif()
+  endforeach()
+  
+  # These are out-of-order in `LLVM_AVAILABLE_LIBS` and should always be last.
+  list(REMOVE_ITEM LLVM_LIBRARIES LLVMMC LLVMCore LLVMSupport)
+  list(APPEND LLVM_LIBRARIES LLVMMC LLVMCore LLVMSupport)
+  
+  target_link_libraries("${target_name}" INTERFACE
+    ${LLVM_LIBRARIES}
+  )
```
Comment (ekilmer) [inline cmake/llvm.cmake:35]: What was wrong with copying what was previously used to set up LLVM? And why are we creating another library `llvm`? There's nothing wrong with using the list of the real upstream LLVM targets in our targets

File: CMakeLists.txt [hunk 7]
```diff
@@ -46,65 +40,31 @@ endif()
...
-find_package(glog CONFIG QUIET)
-if (NOT glog_FOUND)
-  find_package(glog REQUIRED)
-  list(APPEND PROJECT_LIBRARIES glog)
-else()
-  list(APPEND PROJECT_LIBRARIES glog::glog)
-endif()
+find_package(gflags CONFIG REQUIRED HINTS "${_VCPKG_SHARE_DIR}")
+find_package(glog CONFIG REQUIRED HINTS "${_VCPKG_SHARE_DIR}")
+find_llvm(llvm "${_VCPKG_SHARE_DIR}")
```
Comment (ekilmer) [inline CMakeLists.txt:54]: None of these hints should be required, and if they are, then the cxx-common vcpkg packaging should be changed.

What are the issues you are encountering here that makes the `_VCPKG_SHARE_DIR` required for use?

**Review by ekilmer — CHANGES_REQUESTED** [review]
I question a few of these changes. If anything, using vcpkg should make the dependency setup easier and less verbose in the project's CMake.

The other changes look relatively good though.

File: CMakeLists.txt (hunk 3 above)
Comment (pgoodman) [reply CMakeLists.txt:63]: I haven't had issues with Z3 now that we build it in vcpkg; really I could do `find_package(Z3...)` before LLVM.

File: CMakeLists.txt (hunk 4 above)
Comment (pgoodman) [reply CMakeLists.txt:9]: It means the need for additional logic, like fixing up z3 things. It means we can't make guarantees on the llvm version, etc. I would rather we require that we only support vcpkg-based builds.

File: CMakeLists.txt (hunk 5 above)
Comment (pgoodman) [reply CMakeLists.txt:50]: There was a case in IRENE, probably for remill, where I needed this, so I've been copying it around.

File: cmake/llvm.cmake (hunk 6 above)
Comment (pgoodman) [reply cmake/llvm.cmake:35]: the real upstream `LLVM_AVAILABLE_LIBS ` list is out of order, and so linking can lead to issues.

File: CMakeLists.txt (hunk 7 above)
Comment (pgoodman) [reply CMakeLists.txt:54]: I no longer remember, but it may have been remill.

File: CMakeLists.txt (hunk 3 above)
Comment (ekilmer) [reply CMakeLists.txt:63]: Ah yes, you are right. Pretty sure I did fix that issue in cxx-common now.

File: README.md (hunk 1 above)
Comment (pgoodman) [reply README.md:99]: Done in next commit.

File: CMakeLists.txt (hunk 3 above)
Comment (pgoodman) [reply CMakeLists.txt:63]: Done in next commit.

File: CMakeLists.txt (hunk 5 above)
Comment (pgoodman) [reply CMakeLists.txt:50]: Done in next commit.

File: rellicConfig.cmake.in (hunk 2 above)
Comment (pgoodman) [reply rellicConfig.cmake.in]: Addressed in latest commit.

**Review by ekilmer — APPROVED** [review]
If it passes CI, then LGTM!

**Review by frabert — APPROVED** [review]

**Comment by frabert** [comment]
The test that's causing the timeouts is `bitops`, so I suspect it has got something to do with Z3

**Comment by frabert** [comment]
The offending loop seems to be this one: https://github.com/lifting-bits/rellic/blob/739788810b94a659a100686eb6dfb468020eed06/tools/decomp/Decomp.cpp#L256

Something seems to keep modifying stuff

**Pull request merged**